### PR TITLE
AP_Log - replacement interface for logging

### PR DESCRIPTION
**Pull request by peterbarker**
Some discussion here: https://github.com/ArduPilot/ardupilot/pull/6864

@lucasdemarchi wants `ap::Log` rather than AP_Log.

I'm not actually 100% convinced ATM.... one drawback is while these look like static calls, they're really not.  They're just pretending, they really do work on an instance.

### CODE REVIEW

**Comment by peterbarker** [comment]
There are no plans, really.... it's all up for grabs.  I just though I'd formalise the discussion a little with a PR (also providing a place to have that discussion that won't get closed when an only vaguely related PR gets closed)

**Comment by tridge** [comment]
I strongly support renaming DataFlash_Class to AP_Logging. I'd also find AP_Log acceptable, but prefer AP_Logging as clearer (so it doesn't get thought of as a math library).

**Comment by lucasdemarchi** [comment]
> I'm not actually 100% convinced ATM.... one drawback is while these look like static calls, they're really not. They're just pretending, they really do work on an instance.

so... why not do:

``` c++

// AP_Logging.h
namespace ap {
extern AP_Logging logging;
}

// AP_Logging.cpp
namespace ap {
AP_Logging logging;
}

// my_source.cpp
ap::logging.write(...);
```

Or even without namespace. And even without a class, like in plain good C :). Any plans to update this PR?

**Comment by peterbarker** [comment]
On Thu, 21 Sep 2017, Lucas De Marchi wrote:

> // my_source.cpp
> ap::logging.write(...);

I quite like that.  However, there's been some resistance to using c++ 
namespaces, particularly when::they::get::quite::deep.

> Or even without namespace. And even without a class, like in plain good C

logging.write() is pretty close to what we do with 
gcs().send_statustext(...), so that's also worth considering.  Perhaps it 
is too generic a term and we might hit name conflicts with random things 
we try to use, however.

> :). Any plans to update this PR?

I'm hoping for comments, which I just got (thanks!).  I might do an 
alternative branch or two (if not PRs) with the styles you suggest here, 
so we can see what they look writ large(r).

**Comment by OXINARF** [comment]
I'm all for namespaces and, although I understand that nested namespaces are ugly, I don't see where we would have a situation where it really was so bad.

**Comment by dgrat** [comment]
Just for interest:

`    if (time_us == 0) {
        time_us = AP_HAL::micros64();
    }
    float climbrate = baro.get_climb_rate();
    float drift_offset = baro.get_baro_drift_offset();
    float ground_temp = baro.get_ground_temperature();
    struct log_BARO pkt = {
        LOG_PACKET_HEADER_INIT(LOG_BARO_MSG),
        time_us       : time_us,
        altitude      : baro.get_altitude(0),
        pressure      : baro.get_pressure(0),
        temperature   : (int16_t)(baro.get_temperature(0) * 100 + 0.5f),
        climbrate     : climbrate,
        sample_time_ms: baro.get_last_update(0),
        drift_offset  : drift_offset,
        ground_temp   : ground_temp,
    };
    WriteBlock(&pkt, sizeof(pkt));

    if (baro.num_instances() > 1 && baro.healthy(1)) {
        struct log_BARO pkt2 = {
            LOG_PACKET_HEADER_INIT(LOG_BAR2_MSG),
            time_us       : time_us,
            altitude      : baro.get_altitude(1),
            pressure      : baro.get_pressure(1),
            temperature   : (int16_t)(baro.get_temperature(1) * 100 + 0.5f),
            climbrate     : climbrate,
            sample_time_ms: baro.get_last_update(1),
            drift_offset  : drift_offset,
            ground_temp   : ground_temp,
        };
        WriteBlock(&pkt2, sizeof(pkt2));
    }

    if (baro.num_instances() > 2 && baro.healthy(2)) {
        struct log_BARO pkt3 = {
            LOG_PACKET_HEADER_INIT(LOG_BAR3_MSG),
            time_us       : time_us,
            altitude      : baro.get_altitude(2),
            pressure      : baro.get_pressure(2),
            temperature   : (int16_t)(baro.get_temperature(2) * 100 + 0.5f),
            climbrate     : climbrate,
            sample_time_ms: baro.get_last_update(2),
            drift_offset  : drift_offset,
            ground_temp   : ground_temp,
        };
        WriteBlock(&pkt3, sizeof(pkt3));
    }`

You created dozens of function implemented like this. Wouldn't it be possible to define in the device classes a function creating a info tuple. In the Datalogger you grab the type and print the tuple? Doing so would avoid the definition of these structs and one would just need one function in the DataLogger ...

**Comment by peterbarker** [comment]
@dgrat Thanks for the feedback.  Kind of a separate issue to what we're discussing here, but useful nonetheless.

The specific example you've given here is interesting in that it looks trivially factorisable: https://github.com/ArduPilot/ardupilot/pull/7013 .  Do you see advantages in using tuples rather than a traditionally factoring like this?

Note that there's traditionally been a fair bit of resistance to templates as code-size blowout is relatively easy to achieve (not to mention the syntactic soup issues).

**Comment by dgrat** [comment]
@peterbarker 
>  Do you see advantages in using tuples rather than a traditionally factoring like this?

At the first glance, the advantage is that one would _not_ need to define container structs. So most of LogStructure.h can be omitted. 

> Note that there's traditionally been a fair bit of resistance to templates as code-size blowout is relatively easy to achieve (not to mention the syntactic soup issues).

A possible template implementation would be relatively compact and anyway hidden by a namespace. The visible API would include just simple to use functions without obligatory template parameters.

**Comment by peterbarker** [comment]
I have pushed up a new version of this `AP_Logging` branch based on @lucasdemarchi 's suggestion.

I quite like it.

Note the nouning to get `logger` as the object on which to call the logging methods.  Also note the use of a method to get the actual logging object; that makes it work like `gcs()` and (soon) `gps()`.

Thoughts on this new look?

File: libraries/AC_AttitudeControl/ControlMonitor.cpp [hunk 1]
```diff
@@ -41,14 +42,15 @@ void AC_AttitudeControl::control_monitor_update(void)
...
 {
-    DataFlash_Class::instance()->Log_Write("CTRL", "TimeUS,RMSRollP,RMSRollD,RMSPitchP,RMSPitchD,RMSYaw", "Qfffff",
-                                           AP_HAL::micros64(),
-                                           (double)sqrtf(_control_monitor.rms_roll_P),
-                                           (double)sqrtf(_control_monitor.rms_roll_D),
-                                           (double)sqrtf(_control_monitor.rms_pitch_P),
-                                           (double)sqrtf(_control_monitor.rms_pitch_D),
-                                           (double)sqrtf(_control_monitor.rms_yaw));
-
+    AP::logger().write("CTRL",
```
Comment (lucasdemarchi) [inline libraries/AC_AttitudeControl/ControlMonitor.cpp]: I like this. But then maybe the class/library should be AP_Logger rather than AP_Logging?

**Comment by peterbarker** [comment]
On Wed, 25 Oct 2017, Lucas De Marchi wrote:

> I like this. But then maybe the class/library should be AP_Logger rather
> than AP_Logging?

I'm glad I wasn't the only one thinking that :-)

**Comment by amilcarlucas** [comment]
needs a rebase

**Comment by peterbarker** [comment]
Moved this along.

Conversion done using this script:

```
#!/bin/bash

set -e
set -x

pushd libraries
  git mv DataFlash AP_Logger
  pushd AP_Logger
    for i in DataFlash_*; do x=`echo "$i" | perl -pe 's/DataFlash_/AP_Logger_/'`; echo git mv $i $x; done | sh
    pushd examples
      git mv DataFlash_AllTypes AP_Logger_AllTypes
      git mv DataFlash_test AP_Logger_test
      pushd AP_Logger_AllTypes
        for i in DataFlash_*; do x=`echo "$i" | perl -pe 's/DataFlash_/AP_Logger_/'`; echo git mv $i $x; done | sh
      popd
      pushd AP_Logger_test
        for i in DataFlash_*; do x=`echo "$i" | perl -pe 's/DataFlash_/AP_Logger_/'`; echo git mv $i $x; done | sh
      popd
    popd
    git mv DataFlash.h AP_Logger.h
    git mv DataFlash.cpp AP_Logger.cpp
    git mv DFMessageWriter.cpp LoggerMessageWriter.cpp
    git mv DFMessageWriter.h LoggerMessageWriter.h
  popd
popd

git commit -m "DataFlash: rename to AP_Logger"

perl -pe 's/DataFlash/AP_Logger/' -i Tools/ardupilotwaf/ardupilotwaf.py
git commit -m "Tools: ardupilotwaf: DataFlash renamed to AP_Logger" Tools/ardupilotwaf/ardupilotwaf.py

for i in `find . -name \*.cpp -or -name \*.h -and -not -path "*/modules/*"`; do
 if [ ! -f $i ]; then
    continue
 fi
 if [ -h $i ]; then
    continue
 fi
 perl -pe '
      s/DataFlash[.]/logger./g;
      s/DataFlash_Class/AP_Logger/g;
      s/DataFlash/AP_Logger/g;
      s/(rover|plane|copter|sub|tracker).AP_Logger/$1.logger/g;
      s%AP_Logger/logger.h%AP_Logger/AP_Logger.h%g;
      s%DFMessageWriter%LoggerMessageWriter%g;
      s%&AP_Logger%&logger%g;
      s%"logger.h"%"AP_Logger.h"%g;
      s%&logger_%&AP_Logger_%g;
      s%AP_Logger AP_Logger;%AP_Logger logger;%;
      s%AP_Logger\(g.log_bitmask\)%logger(g.log_bitmask)%;
      s%k_param_AP_Logger%k_param_logger%;
      s%AP_Logger\{g.log_bitmask\}%logger{g.log_bitmask}%;
 ' -i $i
done

for i in Ardu* AntennaTracker APMrover2; do 
    perl -pe 's/GOBJECT\(AP_Logger,/GOBJECT\(logger,/' -i $i/Parameters.cpp
done

for i in APMrover2/Rover.h ArduPlane/Plane.h ArduCopter/Copter.h ArduSub/Sub.h AntennaTracker/Tracker.h; do
    perl -pe 's/AP_Logger AP_Logger;/AP_Logger logger;/' -i $i
done

git add -u libraries ArduSub ArduPlane ArduCopter APMrover2 AntennaTracker Tools
git commit -m "GLOBAL: rename DataFlash_Class to AP_Logger"

# grab the AP::logger() method:
git cherry-pick 44e2d1132a

# rename functions
for i in `find . -name \*.cpp -or -name \*.h -and -not -path "*/modules/*"`; do
 if [ ! -f $i ]; then
    continue
 fi
 if [ -h $i ]; then
    continue
 fi
 perl -pe '
      s/logger[.]Log_/logger./g;
      s/AP_Logger::instance\(\)->/AP::logger()./g;
      s/AP::logger\(\)[.]Log_/AP::logger()./g;
      s/->Log_Write/->Write/g;
      s/[.]Log_Write/.Write/g;
      s/ahrs.Write_Home_And_Origin/ahrs.Log_Write_Home_And_Origin/g;
      s/sitl.Write_SIMSTATE/sitl.Log_Write_SIMSTATE/g;
      s/Write_Error/Log_Write_Error/g;
      s/Write_Event/Log_Write_Event/g;
      s/Write_GuidedTarget/Log_Write_GuidedTarget/g;
      s/Log_Log_Write/Log_Write/g;
      s/setVehicle_Startup_Log_Writer/setVehicle_Startup_Writer/g;
 ' -i $i
done

pushd libraries/AP_Logger
for i in *.cpp *.h; do
  perl -pe 's/Log_//g' -i $i
done
popd


git add -u libraries ArduSub ArduPlane ArduCopter APMrover2 AntennaTracker Tools
git commit -m "GLOBAL: use AP::logger() and strip redundant Log_ from methods"
```

**Comment by lucasdemarchi** [comment]
\o/

**Pull request merged**